### PR TITLE
Add state for Mint: tick and Position

### PR DESCRIPTION
**Pull request by mattverse**
This PR mainly serves two purposes:
- Adding State for tick 
- Adding State for position 
in the process of Minting. 

1. Adding State for Tick
We have two state entries for TickInfo now, `liquidityGross` and `liquidityNet`. Gross checks the inclined amount regardless of liquidityDelta being positive or negative, while liquidityNet tracks liquidity with regarding the sign of liquidityDelta. 

2. Adding state for Position 
We add state for position given address, upper tick, lower tick and liquidity In whenever a mint is happening. 

Added tests to verify that state entry is correctly being stored.

### CODE REVIEW

File: x/concentrated-liquidity/lp.go [hunk 1]
```diff
@@ -24,14 +24,28 @@ func (k Keeper) Mint(ctx sdk.Context, poolId uint64, owner sdk.AccAddress, liqui
...
+	// TODO: come back to sdk.Int vs sdk.Dec state & truncation
+	err = k.initOrUpdateTick(ctx, poolId, lowerTick, liquidityIn.TruncateInt(), false)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, nil
+	}
+
+	// TODO: come back to sdk.Int vs sdk.Dec state & truncation
+	err = k.initOrUpdateTick(ctx, poolId, upperTick, liquidityIn.TruncateInt(), true)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, nil
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/lp.go:38]:
```suggestion
		return sdk.Int{}, sdk.Int{}, err
```

File: x/concentrated-liquidity/lp_test.go [hunk 2]
```diff
@@ -31,7 +31,22 @@ func (s *KeeperTestSuite) TestMint() {
...
+	// check tick state
+	lowerTickInfo, err := s.App.ConcentratedLiquidityKeeper.GetTickInfo(s.Ctx, poolId, lowerTick)
+	s.Require().NoError(err)
+	s.Require().Equal(sdk.NewInt(1517), lowerTickInfo.LiquidityGross)
+	s.Require().Equal(sdk.NewInt(1517), lowerTickInfo.LiquidityNet)
+
+	upperTickInfo, err := s.App.ConcentratedLiquidityKeeper.GetTickInfo(s.Ctx, poolId, upperTick)
+	s.Require().NoError(err)
+	s.Require().Equal(sdk.NewInt(1517), upperTickInfo.LiquidityGross)
+	s.Require().Equal(sdk.NewInt(-1517), upperTickInfo.LiquidityNet)
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/lp_test.go:51]: can you explain the math behind getting 1517?

File: x/concentrated-liquidity/position.go [hunk 3]
```diff
@@ -2,36 +2,53 @@ package concentrated_liquidity
...
+	}
 
 	liquidityBefore := position.Liquidity
-	liquidityAfter := liquidityBefore.Add(liquidityDelta)
+	var liquidityAfter sdk.Int
+	if liquidityDelta.IsNegative() {
+		liquidityAfter = liquidityBefore.Sub(liquidityDelta)
+	} else {
+		liquidityAfter = liquidityBefore.Add(liquidityDelta)
+	}
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/position.go:28]: to be clear, this means that no matter what, we are adding liquidity to liquidityAfter, even if we have a negative delta right?

File: x/concentrated-liquidity/tick.go [hunk 4]
```diff
@@ -87,12 +117,23 @@ func (k Keeper) NextInitializedTick(ctx sdk.Context, poolId uint64, tickIndex in
 	return 0, false
 }
 
-func (k Keeper) getTickInfo(ctx sdk.Context, poolId uint64, tickIndex int64) TickInfo {
+// getTickInfo gets tickInfo given poolId and tickIndex. Returns a boolean field that returns true if value is found for given key.
+func (k Keeper) GetTickInfo(ctx sdk.Context, poolId uint64, tickIndex int64) (tickInfo TickInfo, err error) {
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/tick.go:121]: is there a reason we also made this public?

**Review by czarcas7ic — APPROVED** [review]
🥇

File: x/concentrated-liquidity/lp_test.go (hunk 2 above)
Comment (mattverse) [reply x/concentrated-liquidity/lp_test.go:51]: It's the liquidity we put in initially as the testing params

File: x/concentrated-liquidity/tick.go (hunk 4 above)
Comment (mattverse) [reply x/concentrated-liquidity/tick.go:121]: for testing purposes / I don't think we have security vulnerability exposing getter method

File: x/concentrated-liquidity/position.go [hunk 5]
```diff
@@ -2,36 +2,53 @@ package concentrated_liquidity
...
+
 	position.Liquidity = liquidityAfter
 
 	k.setPosition(ctx, poolId, owner, lowerTick, upperTick, position)
+	return nil
 }
 
-// nolint: unused
-func (k Keeper) getPosition(ctx sdk.Context, poolId uint64, owner sdk.AccAddress, lowerTick, upperTick int64) Position {
+func (k Keeper) GetPosition(ctx sdk.Context, poolId uint64, owner sdk.AccAddress, lowerTick, upperTick int64) (position Position, err error) {
```
[the remark this answers is not in the document]
Comment (mattverse) [reply x/concentrated-liquidity/position.go:36]: for testing purposes / I don't think we have security vulnerability exposing getter method

File: x/concentrated-liquidity/lp_test.go (hunk 2 above)
Comment (p0mvn) [reply x/concentrated-liquidity/lp_test.go:51]: Should we add a comment for that?

File: x/concentrated-liquidity/position.go (hunk 3 above)
Comment (mattverse) [reply x/concentrated-liquidity/position.go:28]: So if we have a negative delta, we don't want to subtract it from liquidityBefore, so we do another negative to add the absolute value of liquidityDelta

Comment (p0mvn) [reply x/concentrated-liquidity/position.go:28]: Yeah this seems odd to me. If it's negative, we should be subtracting, right?

File: x/concentrated-liquidity/position.go [hunk 6]
```diff
@@ -2,36 +2,53 @@ package concentrated_liquidity
...
-	var position Position
+	positionStruct := Position{}
 	key := types.KeyPosition(poolId, owner, lowerTick, upperTick)
-	osmoutils.MustGet(store, key, &position)
 
-	return position
+	bz := store.Get(key)
+	if bz == nil {
+		return Position{Liquidity: sdk.ZeroInt()}, nil
+	}
```
Comment (p0mvn) [inline x/concentrated-liquidity/position.go:44]: Can we use `osmoutils.MustGet` here?

File: x/concentrated-liquidity/tick.go [hunk 7]
```diff
@@ -87,12 +117,23 @@ func (k Keeper) NextInitializedTick(ctx sdk.Context, poolId uint64, tickIndex in
...
 	store := ctx.KVStore(k.storeKey)
-	tickInfo := TickInfo{}
+	tickStruct := TickInfo{}
 	key := types.KeyTick(poolId, tickIndex)
-	osmoutils.MustGet(store, key, &tickInfo)
-	return tickInfo
+
+	bz := store.Get(key)
```
Comment (p0mvn) [inline x/concentrated-liquidity/tick.go:126]: `osmoutils.MustGet` if applicable please

**Review by p0mvn — APPROVED** [review]
Would like to know if we should always be subtracting delta and use `osmoutils.MustGet`. Otherwise, LGTM

File: x/concentrated-liquidity/lp_test.go (hunk 2 above)
Comment (mattverse) [reply x/concentrated-liquidity/lp_test.go]: added

File: osmoutils/store_helper.go [hunk 8]
```diff
@@ -124,6 +124,19 @@ func MustGet(store store.KVStore, key []byte, result proto.Message) {
 	}
 }
 
+// GetIfFound gets ket from store by mutating result.
```
Comment (czarcas7ic) [inline osmoutils/store_helper.go:127]:
```suggestion
// GetIfFound gets key from store
```

File: x/concentrated-liquidity/position.go [hunk 9]
```diff
@@ -7,31 +7,44 @@ import (
...
 	store := ctx.KVStore(k.storeKey)
-
-	var position Position
+	positionStruct := Position{}
 	key := types.KeyPosition(poolId, owner, lowerTick, upperTick)
-	osmoutils.MustGet(store, key, &position)
 
-	return position
+	found, err := osmoutils.GetIfFound(store, key, &positionStruct)
+	// return 0 values iuf key has not been initialized
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/position.go:39]:
```suggestion
	// return 0 values if key has not been initialized
```

**Review by czarcas7ic — APPROVED** [review]

**Pull request merged**